tests/e2e/action_rapid_test.go

```diff
@@ -0,0 +1,501 @@
+// deserialized without error.
+// The generators for the various actions are defined in this file, and
+// essentially tell rapid how to build these actions.
+
+func TestActionMarshalling(t *testing.T) {
+	rapid.Check(t, func(t *rapid.T) {
+		action := GetActionGen().Draw(t, "Action")
+		err := MarshalAndUnmarshalAction(action)
+		if err != nil {
+			t.Fatalf("error marshalling and unmarshalling action: %v", err)
+		}
+	})
+}
+
+func MarshalAndUnmarshalAction(action interface{}) error {
+	// wraps the action with a step, since it needs custom unmarshalling that is called by the step unmarshaller
+	step := Step{
+		Action: action,
+	}
+	jsonobj, err := json.Marshal(step)
+	if err != nil {
+		return fmt.Errorf("error marshalling action inside step: %v", err)
+	}
+
+	var got Step
+	err = json.Unmarshal(jsonobj, &got)
+	if err != nil {
+		return fmt.Errorf("error unmarshalling action inside step: %v", err)
+	}
+
+	diff := cmp.Diff(step, got)
+	if diff != "" {
+		return fmt.Errorf("got (-), want (+): %v", diff)
+	}
+
+	return nil
+}
+
+// This needs to be adjusted manually when new actions are added and should
+// include generators for all actions that are mentioned in main.go/runStep.
+func GetActionGen() *rapid.Generator[any] {
+	return rapid.OneOf(
+		GetStartSovereignChainActionGen().AsAny(),
+		GetSubmitLegacyUpgradeProposalActionGen().AsAny(),
+		GetWaitUntilBlockActionGen().AsAny(),
+		GetChangeoverChainActionGen().AsAny(),
+		GetSendTokensActionGen().AsAny(),
+		GetStartChainActionGen().AsAny(),
+		GetSubmitTextProposalActionGen().AsAny(),
+		GetSubmitConsumerAdditionProposalActionGen().AsAny(),
+		GetSubmitConsumerRemovalProposalActionGen().AsAny(),
+		GetSubmitParamChangeProposalActionGen().AsAny(),
+		GetSubmitEquivocationProposalActionGen().AsAny(),
+		GetVoteGovProposalActionGen().AsAny(),
+		GetStartConsumerChainActionGen().AsAny(),
+		GetAddChainToRelayerActionGen().AsAny(),
+		GetAddIbcConnectionActionGen().AsAny(),
+		GetAddIbcChannelActionGen().AsAny(),
+		GetStartRelayerActionGen().AsAny(),
+		GetTransferChannelCompleteActionGen().AsAny(),
+		GetRelayPacketsActionGen().AsAny(),
+		GetRelayRewardPacketsToProviderActionGen().AsAny(),
+		GetDelegateTokensActionGen().AsAny(),
+		GetUnbondTokensActionGen().AsAny(),
+		GetRedelegateTokensActionGen().AsAny(),
+		GetDowntimeSlashActionGen().AsAny(),
+		GetUnjailValidatorActionGen().AsAny(),
+		GetRegisterRepresentativeActionGen().AsAny(),
+		GetDoublesignSlashActionGen().AsAny(),
+		GetAssignConsumerPubKeyActionGen().AsAny(),
+		GetSlashThrottleDequeueActionGen().AsAny(),
+		GetCreateIbcClientsActionGen().AsAny(),
+		CreateCancelUnbondTokensActionGen().AsAny(),
+		CreateLightClientEquivocationAttackActionGen().AsAny(),
+		CreateLightClientAmnesiaAttackActionGen().AsAny(),
+		CreateLightClientLunaticAttackActionGen().AsAny(),
+	)
+}
+
+func CreateSubmitChangeRewardDenomsProposalActionGen() *rapid.Generator[submitChangeRewardDenomsProposalAction] {
```

Comment:
These funcs all share some boilerplate, so I messed around for a sec with using generics to make it more succicnt. Here's what I found:

You can define a generic GetAction func like so,
```go
func GetAction[T any](callback func(t *rapid.T) T) *rapid.Generator[T] {
	return rapid.Custom(callback)
}
```

and call it like so
```go
GetAction[submitChangeRewardDenomsProposalAction](
	func(t *rapid.T) submitChangeRewardDenomsProposalAction {
		return submitChangeRewardDenomsProposalAction{
			From:    GetValidatorIDGen().Draw(t, "From"),
			Deposit: rapid.Uint().Draw(t, "Deposit"),
			Denom:   rapid.String().Draw(t, "Denom"),
		}
	},
).AsAny(),
```

This reduces a couple lines of code for each action, in that you don't have to define a new func for each one. 

Nbd if you don't want to use this idea, just a bit of golang exploring on my part

Reply:
Thanks for exploring this! I think I personally like the current way a bit more, even if its slightly more verbose.